File: src/main/kotlin/com/geumjulee/meow_diary/service/CatService.kt
package com.geumjulee.meow_diary.service

import com.geumjulee.meow_diary.dto.CatCreateRequest
import com.geumjulee.meow_diary.dto.CatResponse
import com.geumjulee.meow_diary.dto.CatUpdateRequest
import com.geumjulee.meow_diary.entity.Cat
import com.geumjulee.meow_diary.repository.CatRepository
import com.geumjulee.meow_diary.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class CatService(
    private val catRepository: CatRepository,
    private val userRepository: UserRepository
) {
    
    fun createCat(userId: Long, request: CatCreateRequest): CatResponse {
        val user = userRepository.findById(userId)
            .orElseThrow { IllegalArgumentException("사용자를 찾을 수 없습니다") }
        
        val cat = Cat().apply {
            name = request.name
            breed = request.breed
            color = request.color
            birthDate = request.birthDate
            weight = request.weight
            gender = request.gender
            isNeutered = request.isNeutered
            description = request.description
            this.user = user
        }
        
        val savedCat = catRepository.save(cat)
        return CatResponse.from(savedCat)
    }
    
    @Transactional(readOnly = true)
    fun getCatById(id: Long): CatResponse {
        val cat = catRepository.findById(id)
            .orElseThrow { IllegalArgumentException("고양이를 찾을 수 없습니다") }
        return CatResponse.from(cat)
    }
    
    @Transactional(readOnly = true)
    fun getCatsByUserId(userId: Long): List<CatResponse> {
        return catRepository.findByUserId(userId).map { CatResponse.from(it) }
    }
    
    fun updateCat(id: Long, request: CatUpdateRequest): CatResponse {
        val cat = catRepository.findById(id)
            .orElseThrow { IllegalArgumentException("고양이를 찾을 수 없습니다") }
        
        request.name?.let { cat.name = it }
        request.breed?.let { cat.breed = it }
        request.color?.let { cat.color = it }
        request.birthDate?.let { cat.birthDate = it }
        request.weight?.let { cat.weight = it }
        request.gender?.let { cat.gender = it }
        request.isNeutered?.let { cat.isNeutered = it }
        request.profileImageUrl?.let { cat.profileImageUrl = it }
        request.description?.let { cat.description = it }
        
        val updatedCat = catRepository.save(cat)
        return CatResponse.from(updatedCat)
    }
    
    fun deleteCat(id: Long) {
        val cat = catRepository.findById(id)
            .orElseThrow { IllegalArgumentException("고양이를 찾을 수 없습니다") }
        
        catRepository.delete(cat)
    }
    
    @Transactional(readOnly = true)
    fun searchCatsByName(name: String): List<CatResponse> {
        return catRepository.findByNameContainingIgnoreCase(name).map { CatResponse.from(it) }
    }
    
    @Transactional(readOnly = true)
    fun getCatsByBreed(breed: String): List<CatResponse> {
        return catRepository.findByBreed(breed).map { CatResponse.from(it) }
    }
    
    @Transactional(readOnly = true)
    fun getCatsByGender(gender: String): List<CatResponse> {
        return catRepository.findByGender(gender).map { CatResponse.from(it) }
    }
    
    @Transactional(readOnly = true)
    fun getCatsByNeuteredStatus(isNeutered: Boolean): List<CatResponse> {
        return catRepository.findByNeuteredStatus(isNeutered).map { CatResponse.from(it) }
    }
} 

File: flutter_app/lib/services/cat_service.dart
import 'dart:io';
import 'package:dio/dio.dart';
import '../models/cat.dart';
import 'api_service.dart';

class CatService {
  final ApiService _apiService = ApiService();
  
  Future<List<Cat>> getCatsByUserId(int userId) async {
    try {
      final response = await _apiService.get('/cats/user/$userId');
      final List<dynamic> catsJson = response.data;
      return catsJson.map((json) => Cat.fromJson(json)).toList();
    } catch (e) {
      throw Exception('고양이 목록을 가져오는데 실패했습니다: $e');
    }
  }
  
  Future<Cat> getCatById(int id) async {
    try {
      final response = await _apiService.get('/cats/$id');
      return Cat.fromJson(response.data);
    } catch (e) {
      throw Exception('고양이 정보를 가져오는데 실패했습니다: $e');
    }
  }
  
  Future<Cat> createCat(int userId, CatCreateRequest request, {File? imageFile}) async {
    try {
      if (imageFile != null) {
        // 이미지와 함께 고양이 등록
        final formData = FormData.fromMap({
          'userId': userId.toString(),
          'name': request.name,
          'breed': request.breed ?? '',
          'color': request.color ?? '',
          'birthDate': request.birthDate?.toIso8601String() ?? '',
          'weight': request.weight?.toString() ?? '',
          'gender': request.gender?.name ?? '',
          'isNeutered': request.isNeutered.toString(),
          'description': request.description ?? '',
          'image': await MultipartFile.fromFile(
            imageFile.path,
            filename: 'cat_image.jpg',
          ),
        });
        
        final response = await _apiService.postFormData('/cats', formData);
        return Cat.fromJson(response.data);
      } else {
        // JSON으로 고양이 등록
        final response = await _apiService.post('/cats/json?userId=$userId', 
            data: request.toJson());
        return Cat.fromJson(response.data);
      }
    } catch (e) {
      throw Exception('고양이 등록에 실패했습니다: $e');
    }
  }
  
  Future<Cat> updateCat(int id, CatUpdateRequest request, {File? imageFile}) async {
    try {
      if (imageFile != null) {
        // 이미지와 함께 고양이 수정
        final formData = FormData.fromMap({
          'name': request.name,
          'breed': request.breed,
          'color': request.color,
          'birthDate': request.birthDate?.toIso8601String(),
          'weight': request.weight?.toString(),
          'gender': request.gender?.name,
          'isNeutered': request.isNeutered?.toString(),
          'description': request.description,
          'image': await MultipartFile.fromFile(
            imageFile.path,
            filename: 'cat_image.jpg',
          ),
        });
        
        final response = await _apiService.postFormData('/cats/$id', formData);
        return Cat.fromJson(response.data);
      } else {
        // JSON으로 고양이 수정
        final response = await _apiService.put('/cats/$id', data: request.toJson());
        return Cat.fromJson(response.data);
      }
    } catch (e) {
      throw Exception('고양이 정보 수정에 실패했습니다: $e');
    }
  }
  
  Future<void> deleteCat(int id) async {
    try {
      await _apiService.delete('/cats/$id');
    } catch (e) {
      throw Exception('고양이 삭제에 실패했습니다: $e');
    }
  }
}